### PR TITLE
Cloak should be detected by horizontal range.

### LINKED ISSUE #20973: Stealth detection on flying aircraft is not the full range
bug

In TS and 3rd party mods we have stealth detection air unit (JumpJet for TS), but the stealth detection circle on flying aircraft is actually smaller than it should be.

### PR DESCRIPTION
**Pull request by dnqbob**
or it doesn't match the circle shown by `RenderDetectionCircle` in TS mod. Meanwhile, this will mess up aircraft stealth detection when they changing height.

Fix https://github.com/OpenRA/OpenRA/issues/20973

### CODE REVIEW

**Comment by dnqbob** [comment]
Fix #20973

**Comment by PunkPun** [comment]
I don't really want to move forward with this PR without the detection circle changes https://github.com/OpenRA/OpenRA/issues/20973#issuecomment-1651076305

**Comment by dnqbob** [comment]
Added

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 1]
```diff
@@ -71,7 +74,7 @@ IEnumerable<IRenderable> RenderCircle(Actor self, DetectionCircleVisibility visi
 				yield break;
 
 			yield return new DetectionCircleAnnotationRenderable(
-				self.CenterPosition,
+				self.CenterPosition - new WVec(WDist.Zero, WDist.Zero, info.RenderOnGround ? self.World.Map.DistanceAboveTerrain(self.CenterPosition) : WDist.Zero),
```
Comment (PunkPun) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]:
```suggestion
				position,
```

And then above add
```csharp
var position = self.CenterPosition;
if (info.RenderOnGround)
	position -= new WVec(0, 0, self.World.Map.DistanceAboveTerrain(position).Length);
```

This should be better style

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: Fixed

**Comment by PunkPun** [comment]
Testing more, the cylindrical detection still feels weird with high detection range units (f.e. mobile sensor array).

Though it's an issue only for isometric mods or mods with a lot of stealth aircraft

**Comment by penev92** [comment]
Weird how?

**Comment by PunkPun** [comment]
Detecting units which are outside the detection circle, or not detecting them when they are inside

**Comment by dnqbob** [comment]
Actually in original TS and RA2, it is like so.

**Comment by PunkPun** [comment]
I can't find any references to range circles in those games

**Comment by dnqbob** [comment]
Sensor Array has one in TS.

**Review by RoosterDragon — APPROVED** [review]
@PunkPun couldn't tell if you still had a concrete objection here. But reading the linked issue and background these changes seem conceptually reasonable to me, and code is fine too.

**Comment by penev92** [comment]
FWIW I believe we agreed (on Discord) that while it makes more sense to have a spherical detection area, it is much better for readability to go for cylindrical detection without a height limit (because otherwise it gets *very* confusing in a 2D RTS).

P.S.: I guess "agreed" was premature.

Here are my thoughts:
> tldr IMO we should take the cylindrical detection (however weird it may be) and add some more visual feedback that it is indeed cylindrical
if we could do that for a sphere then that'd be best, but other than drawing the sphere I don't see how we could

https://discord.com/channels/153649279762694144/388282819371204608/1144712092205731942

**Comment by dnqbob** [comment]
It is sad that I cannot access the Discord to see your discussion, but I am going to say a sphere detection only makes sense for a true 3D rts, considering in OpenRA the aircraft's fly attitude and subterrean unit's underground attitude are not visually correct.

**Comment by PunkPun** [comment]
We may want a detection circle that bends to terrain. Perhaps looking something like the capture circles from ZeroSpace 

IMO spherical detection is better represented by a range circle at the centre of the actor, basically what we have currently. And while cylindrical detection does make more sence, it is really badly represented by our current detection circles.

**Comment by dnqbob** [comment]
A circle bends to ground/casts on terrain requires you game based on a accurate 3D visual, otherwise the result won't be satisfying.

Moreover, I will mention the subterrean unit again, with detection range loss they won't be detected correctly, whether in game or in visual. In this case, using sphere detection means the detection function of TS is fundamentally broken.

**Comment by penev92** [comment]
More from Discord:
Keep the range circle at the actor's altitude (whether ground or air)
> use something like  #18400  to project the range circle downwards to the ground if needed (and maybe upwards, to make it abundantly clear that it's a cylinder?)
> upwards can be similar to what ZeroSpace has for the neutral tower capture ranges+indicators

Despite spherical detection being the more logical one, it is going to be confusing and problematic in a 2D/2.5D RTS game. There will always be issues with not detecting enemies at max range on different altitudes without that being clear to the player. This is why I vote for the cylinder - less logical, but has way better readability for the players.

Of course both will apparently need additional visual aid, as discussed. And while I can see us adding visual aid as mentioned above to show that the area extends upwards and downwards to create a cylinder, I can't imagine how we could do that for a sphere.


P.S.: Actually, can we make the detection configurable? `DetectionRangeType: Spherical/Cylindrical` ? 🤔 This will let mods that do have a use-case for the realism of the sphere to still have it as an option. The range indicator is going to be a separate trait anyway, so is already configurable.

**Comment by dnqbob** [comment]
If we really want sphere detection, I would recommend we take both and make it optional for player.

**Comment by PunkPun** [comment]
I think we want cylindrical detection. It's just that our current UI only supports spherical

**Comment by PunkPun** [comment]
Currently things that are outside the detection circle will never be detected, but things which are inside might, or may not depending on height delta. But with cylindrical things that are outside the circle will also be detected, which looks extremelly jarring

**Comment by dnqbob** [comment]
> Actually, can we make the detection configurable? DetectionRangeType: Spherical/Cylindrical ? 

orca
 Tried, on the second thought I just drop it. The detection range of spherical detection type is just ridiculous on height change (AA detection range is about twice to AG detection range), I don't see any need for us to use that.

In this case, you can see how subterrean unit is escaping from the detection. In another word, if we want to finish TS, cylindrical detection is critical.

**Comment by dnqbob** [comment]
> But with cylindrical things that are outside the circle will also be detected, which looks extremelly jarring

It is not the fault of detection circle, it is we don't have a proper position indicated to render the actual aircraft position on the ground level when they are flying, so do subterrean units.

ta1

**Review by RoosterDragon — APPROVED** [review]
I am still supportive of this change.

**Comment by abcdefg30** [comment]
I also don't understand the problem here. We can display the detection circle on the ground using the new setting introduced in this PR. @dnqbob can you please rebase this PR and resolve the merge conflicts?

**Comment by PunkPun** [comment]
The ground isn't flat though, so the detection circle becomes very misleading

**Comment by abcdefg30** [comment]
Isn't that the same with all our circles at the moment?
